tests/unit/steps/test_plan.py

```diff
@@ -1575,3 +1576,66 @@ def test_create_upgrade_plan_failed():
 
     with pytest.raises(Exception, match="test"):
         cou_plan._create_upgrade_group([app], "victoria", "test", False)
+
+
+@pytest.mark.parametrize(
+    "cli_force, empty_hypervisors, expected_logs",
+    [
+        (
+            True,
+            {},
+            [
+                (
+                    "cou.steps.plan",
+                    logging.INFO,
+                    "Selected all hypervisors: nova-compute/0, nova-compute/1, nova-compute/2",
+                )
+            ],
+        ),
+        (
+            False,
+            {1},
+            [
+                (
+                    "cou.steps.plan",
+                    logging.INFO,
+                    "Found non-empty hypervisors: nova-compute/0, nova-compute/2",
+                ),
+                ("cou.steps.plan", logging.INFO, "Selected hypervisors: nova-compute/1"),
+            ],
+        ),
+    ],
+)
+@pytest.mark.asyncio
+@patch("cou.steps.plan._get_nova_compute_units_and_machines")
+@patch("cou.steps.plan.get_empty_hypervisors")
+async def test_log_upgradable_hypervisors(
```

Comment:
I ended up rewriting the test and creating a separate config for it, for better flexibility. I think the test should focus on testing the logging logic, instead of the plan. Using a customized config can make it easier to control arguments such as cli_force.